state-sync/src/coordinator.rs

```diff
@@ -400,194 +318,265 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
-            if let Some(sync_request) = self.sync_request.take() {
-                Self::send_sync_req_callback(sync_request, Ok(()))?;
+                    .local_epoch(local_epoch));
+                Self::send_initialization_callback(initialization_listener)?;
             }
         }
 
-        let initialization_complete = self
-            .initialization_listener
-            .as_ref()
-            .map_or(false, |_| self.is_initialized());
-        if initialization_complete {
-            info!(LogSchema::event_log(LogEntry::Waypoint, LogEvent::Complete)
-                .local_li_version(committed_version)
-                .local_synced_version(synced_version)
-                .local_epoch(local_epoch));
-            if let Some(listener) = self.initialization_listener.take() {
-                Self::send_initialization_callback(listener, Ok(()))?;
+        // Check if we're now at the sync request target
+        if let Some(sync_request) = self.sync_request.as_ref() {
+            let sync_target_version = sync_request.target.ledger_info().version();
+            if synced_version > sync_target_version {
+                return Err(Error::SyncedBeyondTarget(
+                    synced_version,
+                    sync_target_version,
+                ));
+            }
+            if synced_version == sync_target_version {
+                debug!(
+                    LogSchema::event_log(LogEntry::SyncRequest, LogEvent::Complete)
+                        .local_li_version(committed_version)
+                        .local_synced_version(synced_version)
+                        .local_epoch(local_epoch)
+                );
+                counters::SYNC_REQUEST_RESULT
+                    .with_label_values(&[counters::COMPLETE_LABEL])
+                    .inc();
+                if let Some(sync_request) = self.sync_request.take() {
+                    Self::send_sync_req_callback(sync_request, Ok(()))?;
+                }
             }
         }
+
         Ok(())
     }
 
-    fn get_state(&mut self, callback: oneshot::Sender<SyncState>) {
-        if let Err(e) = self.sync_state_with_local_storage() {
-            error!(
-                "[state sync] failed to sync with local storage for get_state request: {:?}",
-                e
-            );
+    /// Notifies mempool that transactions have been committed.
+    async fn notify_mempool_of_committed_transactions(
+        &mut self,
+        committed_transactions: Vec<Transaction>,
+    ) -> Result<(), Error> {
+        // Get all user transactions from committed transactions
+        let mut user_transactions = vec![];
+        for transaction in committed_transactions {
+            if let Transaction::UserTransaction(signed_txn) = transaction {
+                user_transactions.push(CommittedTransaction {
+                    sender: signed_txn.sender(),
+                    sequence_number: signed_txn.sequence_number(),
+                });
+            }
```

Comment:
It's hard for me to tell what you did and what you moved, but FYI you can do this through a map if you're doing something like this.
```
let user_transactions :Vec<_> = committed_transactions.iter().map_filter(|transaction| {
  match transaction {
   UserTransaction: => Some(...)
  _ => None
 }
}).collect();
```

Reply:
Copy pasta -- fixed 😄